developer/src/server/src/index.ts

```diff
@@ -130,4 +126,68 @@ export async function run() {
     /* Load the tray icon */
     tray.start(configuration.port, configuration.ngrokEndpoint);
   }
+}
+
+function getRunningInstancePid(pidFilename: string) {
+  try {
+    return fs.readFileSync(pidFilename, 'utf-8');
+  } catch(err) {
+    return null;
+  }
+}
+
+function writeLockFile() {
+  const lockFilename = configuration.lockFilename.replaceAll(/[\\\/]/g, path.sep);
+  const pidFilename = configuration.pidFilename.replaceAll(/[\\\/]/g, path.sep);
+
+  // console.debug(`Testing existence of ${lockFilename}`);
+  if(fs.existsSync(lockFilename)) {
+    // If that fails, we'll assume the previous process died and delete the lockfile.
+    try {
+      // console.debug(`Attempting to delete.`);
+      fs.unlinkSync(lockFilename);
+      if(fs.existsSync(lockFilename)) {
+        // console.debug(`Unlink silently failed`);
+        console.warn(`Lock file cannot be deleted, probably owned by ${getRunningInstancePid(pidFilename) ?? 'unknown instance'}`);
+        return false;
+      }
+    } catch(err) {
+      // assume that it is actually locked!
+      console.warn(`Lock file cannot be deleted, probably owned by ${getRunningInstancePid(pidFilename) ?? 'unknown instance'}\n  ${err}`);
+      return false;
+    }
+  }
+
+  // console.debug('Creating lock file');
+  let lockFileDescriptor: number = null;
+  try {
+    lockFileDescriptor = fs.openSync(lockFilename, 'w');
+  } catch(err) {
+    console.warn(`Lock file could not be created, assuming another instance already running, probably ${getRunningInstancePid(pidFilename) ?? 'unknown instance'}\n  ${err}`);
+    return false;
+  }
+  // console.debug(`Lock file descriptor: ${lockFileDescriptor}, pid ${process.pid.toString()}`);
+
+  fs.writeFileSync(pidFilename, process.pid.toString());
+  fs.writeFileSync(lockFileDescriptor, process.pid.toString());
+
+  process.on('exit', () => {
+    if(fs.existsSync(pidFilename)) {
+      fs.unlinkSync(pidFilename);
+    }
+    if(lockFileDescriptor) {
+      fs.closeSync(lockFileDescriptor);
+      lockFileDescriptor = null;
+      try {
+        fs.unlinkSync(lockFilename);
+      } catch(err) {
+        // This can happen if an unlink call is made on a file with an open
+        // descriptor, such as happens above; when the file is closed it is
+        // automatically deleted. On Windows at least.
+        //
+        // console.debug(`Did not unlink ${lockFilename}, probably already deleted`);
```

Comment:
Does this need to return false?

Reply:
no, this `process.on('exit')` closure doesn't return any value.